flink-tests/src/test/java/org/apache/flink/test/runtime/leaderelection/ZooKeeperLeaderElectionITCase.java

```diff
@@ -121,38 +145,39 @@ public void testJobExecutionOnClusterWithLeaderChange() throws Exception {
                 final DispatcherGateway leaderDispatcherGateway =
                         getNextLeadingDispatcherGateway(
                                 miniCluster, previousLeaderAddress, timeout);
+                // Make sure resource manager has also changed leadership.
+                resourceManagerLeaderFutures[i].get();
```

Comment:
Did the test become unstable because of your previous change? If yes, what is causing the instability? I assume it has something to do that we do the leader election for the `Dispatcher` and the `ResourceManager` individually. Isn't this also a problem for Flink in general if we have two processes where one has the leading `ResourceManager` and the other the leading `Dispatcher`? cc @xintongsong this might have been introduced with FLINK-21667.

It seems that we have the assumption in the `DispatcherResourceManagerComponent` that we need a `ResourceManager` to deregister the application. Before, I think it worked because the `ResourceManager` would also do the deregistering w/o having the leadership. One way to solve this problem properly, is to do a single leader election for the JobManager process.

Reply:
Thanks for creating an issue to track this! This seems to be a bigger issue than I initially thought 🤔 

This instability came from asserting that `DispatcherResourceManagerComponent` shutdownFuture doesn't complete exceptionally.

![Screen Shot 2021-08-30 at 8 21 58 AM](https://user-images.githubusercontent.com/299781/131294525-743dbb10-20cc-4437-9768-efb2ceb1828f.png)